controllers/trainingjob/trainingjob_controller.go

```diff
@@ -18,411 +18,347 @@ package trainingjob
-			return RequeueAfterInterval(r.PollInterval, awsErr)
-		} else if r.isSageMaker429Response(awsErr) {
-			log.Info("SageMaker rate limit exceeded, will retry", "err", awsErr)
-			return RequeueAfterInterval(r.PollInterval, awsErr)
+// Clean up any deployments artifacts, then removes the finalizer.
+func (r *Reconciler) cleanupAndRemoveFinalizer(ctx reconcileRequestContext) error {
+	var err error
+
+	if controllers.HasDeletionTimestamp(ctx.TrainingJob.ObjectMeta) {
+		ctx.TrainingJob.ObjectMeta.Finalizers = controllers.RemoveString(ctx.TrainingJob.ObjectMeta.Finalizers, controllers.SageMakerResourceFinalizerName)
+		if err = r.Update(ctx, ctx.TrainingJob); err != nil {
+			return errors.Wrap(err, "Failed to remove finalizer")
 		}
+		ctx.Log.Info("Finalizer has been removed")
+	}
+
+	return nil
+}
+
+// If this function returns an error, the status update has failed, and the reconciler should always requeue.
+// This prevents the case where a terminal status fails to persist to the Kubernetes datastore yet we stop
+// reconciling and thus leave the job in an unfinished state.
+func (r *Reconciler) updateStatus(ctx reconcileRequestContext, trainingJobPrimaryStatus string) error {
+	return r.updateStatusWithAdditional(ctx, trainingJobPrimaryStatus, "", "")
+}
+
+func (r *Reconciler) updateBothStatus(ctx reconcileRequestContext, trainingJobPrimaryStatus, trainingJobSecondaryStatus string) error {
+	return r.updateStatusWithAdditional(ctx, trainingJobPrimaryStatus, trainingJobSecondaryStatus, "")
+}
+
+func (r *Reconciler) updateStatusAndReturnError(ctx reconcileRequestContext, trainingJobPrimaryStatus string, reconcileErr error) error {
+	if err := r.updateBothStatusAndReturnError(ctx, trainingJobPrimaryStatus, "", reconcileErr); err != nil {
+		return errors.Wrapf(reconcileErr, "Unable to update status with error. Status failure was caused by: '%s'", err.Error())
 	}
+	return reconcileErr
+}
 
-	log.Error(err, "Handling unrecoverable sagemaker API error")
+func (r *Reconciler) updateBothStatusAndReturnError(ctx reconcileRequestContext, trainingJobPrimaryStatus, trainingJobSecondaryStatus string, reconcileErr error) error {
+	if err := r.updateStatusWithAdditional(ctx, trainingJobPrimaryStatus, trainingJobSecondaryStatus, reconcileErr.Error()); err != nil {
+		return errors.Wrapf(reconcileErr, "Unable to update status with error. Status failure was caused by: '%s'", err.Error())
+	}
+	return reconcileErr
+}
 
-	etcdUpdateErr := r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-		SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-		TrainingJobStatus:        string(sagemaker.TrainingJobStatusFailed),
-		Additional:               err.Error(),
-		LastCheckTime:            Now(),
-		CloudWatchLogUrl:         cwLogUrl,
-	})
+func (r *Reconciler) updateStatusWithAdditional(ctx reconcileRequestContext, trainingJobPrimaryStatus, trainingJobSecondaryStatus, additional string) error {
```

Comment:
This is where I wish Go had method overloading, because in some cases I think it's beneficial to be able to define both statuses, but others we hard code an empty secondary status. This was my attempt at giving the option, however I do understand it adds complexity.